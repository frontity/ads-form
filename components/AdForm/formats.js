import { flatten } from 'lodash';

export const postLoadFormat = ({ ads, slots = [] }) => {
  const adsByName = ads.fills.reduce((byName, ad) => {
    byName[ad.name] = { ...ad };
    return byName;
  }, {});

  const adNames = ads.fills.map(({ name }) => name);

  const dividedSlots = slots.reduce(
    (divided, slot) =>
      divided.concat(
        slot.names.map(name => {
          const { names: _, ...rest } = slot;
          return { names: [name], ...rest };
        }),
      ),
    [],
  );

  const slotsByAdName = {};
  const noAdSlots = [];

  dividedSlots.forEach(slot => {
    const [name] = slot.names;
    if (adNames.includes(name)) {
      slotsByAdName[name] = slotsByAdName[name] || [];
      slotsByAdName[name].push(slot);
    } else {
      noAdSlots.push(slot);
    }
  });

  Object.entries(slotsByAdName).forEach(([adName, adSlots]) => {
    adsByName[adName].positions = adSlots.map(({ rules, position }) => {
      const items = rules.item ? rules.item.map(({ type }) => type) : [];
      const type = items.some(item =>
        ['home', 'tag', 'category'].includes(item),
      )
        ? 'list'
        : 'single';

      return { type, items, position };
    });
  });

  console.log({ adsByName });

  return { ads: { fills: Object.values(adsByName) }, slots: noAdSlots };
};

export const preSaveFormat = (values, originalValues) => {
  const { ads, slots } = values;
  const { ads: _ads, slots: _slots, ...others } = originalValues;

  const adSlots = flatten(
    ads.fills.map(({ name, positions }) =>
      positions.map(({ items, position }) => ({
        position,
        names: [name],
        rules: { item: items.map(type => ({ type })) },
      })),
    ),
  );

  return {
    ads: { fills: ads.fills.map(({ positions, ...rest }) => ({ ...rest })) },
    slots: slots.concat(adSlots),
    ...others,
  };
};
