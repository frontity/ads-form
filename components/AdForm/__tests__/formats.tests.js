import { postLoadFormat, preSaveFormat } from '../formats';

const databaseValues = {
  ads: {
    fills: [
      {
        name: 'ad_1',
        height: '600',
        positions: [{ type: 'single', items: ['post', 'page'], position: 2 }],
        type: 'doubleclick',
        slot: '/12345678/roba_doble',
        width: '300',
      },
      {
        name: 'ad_3',
        height: '250',
        positions: [{ type: 'single', items: ['post', 'page'], position: 2 }],
        type: 'doubleclick',
        slot: '/12345678/ad_roba',
        width: '300',
      },
    ],
  },
  slots: [
    {
      names: ['iframe'],
      position: 2,
      rules: {
        item: [{ type: 'post' }, { type: 'page' }],
      },
    },
    {
      names: ['ad_1'],
      position: 1,
      rules: {
        item: [{ type: 'post' }, { type: 'page' }, { type: 'media' }],
      },
    },
    {
      names: ['ad_3'],
      position: 2,
      rules: {
        item: [{ type: 'post' }, { type: 'page' }],
      },
    },
  ],
};

const formValues = {
  ads: {
    fills: [
      {
        name: 'ad_1',
        height: '600',
        positions: [
          { type: 'single', items: ['post', 'page', 'media'], position: 1 },
        ],
        type: 'doubleclick',
        slot: '/12345678/roba_doble',
        width: '300',
      },
      {
        name: 'ad_3',
        height: '250',
        positions: [{ type: 'single', items: ['post', 'page'], position: 2 }],
        type: 'doubleclick',
        slot: '/12345678/ad_roba',
        width: '300',
      },
    ],
  },
  slots: [
    {
      names: ['iframe'],
      position: 2,
      rules: {
        item: [{ type: 'post' }, { type: 'page' }],
      },
    },
  ],
};

describe('formats › postLoadFormat', () => {
  test('works as expected', () => {
    expect(postLoadFormat(databaseValues)).toMatchSnapshot();
  });
});

describe('formats › preSaveFormat', () => {
  test('works as expected', () => {
    expect(preSaveFormat(formValues, databaseValues)).toMatchSnapshot();
  });
});
