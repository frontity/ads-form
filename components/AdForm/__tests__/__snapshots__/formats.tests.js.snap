// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formats › postLoadFormat works as expected 1`] = `
Object {
  "ads": Object {
    "fills": Array [
      Object {
        "height": "600",
        "name": "ad_1",
        "positions": Array [
          Object {
            "items": Array [
              "post",
              "page",
            ],
            "position": "some_theme_position",
            "type": "single",
          },
          Object {
            "items": Array [
              "media",
            ],
            "position": 1,
            "type": "media",
          },
        ],
        "slot": "/12345678/roba_doble",
        "type": "doubleclick",
        "width": "300",
      },
      Object {
        "height": "250",
        "name": "ad_3",
        "positions": Array [
          Object {
            "items": Array [
              "post",
              "page",
            ],
            "position": "some_theme_position",
            "type": "single",
          },
        ],
        "slot": "/12345678/ad_roba",
        "type": "doubleclick",
        "width": "300",
      },
    ],
  },
  "slots": Array [
    Object {
      "names": Array [
        "iframe",
      ],
      "position": "some_theme_position",
      "rules": Object {
        "item": Array [
          Object {
            "type": "post",
          },
          Object {
            "type": "page",
          },
        ],
      },
    },
  ],
}
`;

exports[`formats › preSaveFormat works as expected 1`] = `
Object {
  "ads": Object {
    "fills": Array [
      Object {
        "height": "600",
        "name": "ad_1",
        "slot": "/12345678/roba_doble",
        "type": "doubleclick",
        "width": "300",
      },
      Object {
        "height": "250",
        "name": "ad_3",
        "slot": "/12345678/ad_roba",
        "type": "doubleclick",
        "width": "300",
      },
    ],
    "settings": Object {},
  },
  "slots": Array [
    Object {
      "names": Array [
        "iframe",
      ],
      "position": "some_theme_position",
      "rules": Object {
        "item": Array [
          Object {
            "type": "post",
          },
          Object {
            "type": "page",
          },
        ],
      },
    },
    Object {
      "names": Array [
        "ad_1",
      ],
      "position": "some_theme_position",
      "rules": Object {
        "item": Array [
          Object {
            "type": "post",
          },
          Object {
            "type": "page",
          },
        ],
      },
    },
    Object {
      "names": Array [
        "ad_1",
      ],
      "position": "some_theme_position",
      "rules": Object {
        "item": Array [
          Object {
            "type": "media",
          },
        ],
      },
    },
    Object {
      "names": Array [
        "ad_3",
      ],
      "position": "some_theme_position",
      "rules": Object {
        "item": Array [
          Object {
            "type": "post",
          },
          Object {
            "type": "page",
          },
        ],
      },
    },
  ],
}
`;
